File: editorial/ruby/prob3-6-3.rb
def merge_sort(arr)
  # 長さが 1 の場合、すでにソートされているので何もしない
  return arr if arr.length <= 1

  # 2 つに分割した後、小さい配列をソート
  mid = arr.length / 2
  left_half = merge_sort(arr[0...mid])
  right_half = merge_sort(arr[mid..-1])

  # 以下が Merge 操作となります。
  merge(left_half, right_half)
end

def merge(left, right)
  result = []
  until left.empty? || right.empty?
    if left.first <= right.first
      result << left.shift
    else
      result << right.shift
    end
  end
  result + left + right
end

# 以下、メイン部分
n = gets.to_i
a = gets.split.map(&:to_i)

# マージソート → 答えの出力
answer = merge_sort(a)
puts answer.join(' ')

File: math_algorithm/editorial/go/prob3-6-3.go
package main

import (
	"fmt"
)

func mergeSort(arr []int) []int {
	// 長さが 1 の場合、すでにソートされているので何もしない
	if len(arr) <= 1 {
		return arr
	}

	// 2 つに分割した後、小さい配列をソート
	mid := len(arr) / 2
	leftHalf := mergeSort(arr[:mid])
	rightHalf := mergeSort(arr[mid:])

	// 以下が Merge 操作となります。
	return merge(leftHalf, rightHalf)
}

func merge(left, right []int) []int {
	result := make([]int, 0)
	for len(left) > 0 && len(right) > 0 {
		if left[0] <= right[0] {
			result = append(result, left[0])
			left = left[1:]
		} else {
			result = append(result, right[0])
			right = right[1:]
		}
	}
	result = append(result, left...)
	result = append(result, right...)
	return result
}

func main() {
	var n int
	fmt.Scan(&n)
	a := make([]int, n)
	for i := 0; i < n; i++ {
		fmt.Scan(&a[i])
	}

	// マージソート → 答えの出力
	answer := mergeSort(a)
	for _, val := range answer {
		fmt.Printf("%d ", val)
	}
	fmt.Println()
}
